### PR TITLE
Add bootheader xmcd

### PR DESCRIPTION
**Pull request by trunghieulenxp**
**This PR add xmcd bootheader for RT1170.** 

Currently, only DCD bootheader was supported to configure the SDRAM. On IMX RT1170, XMCD can be used as an alternative boot header to DCD. XMCD is more advanced than DCD and enhances SDRAM access speed. This can be benefit for SDRAM access application.

We have benchmarked the performance with Pixel Processing Pipeline (PXP) module.  When using the XMCD boot header, PXP is 13 ms faster compared to DCD bootheader.

As the consequence, the different between DCD and XMCD as bootheader impacts the display pipeline performance.

### CODE REVIEW

**Comment by trunghieulenxp** [comment]
Hi, Could someone please review this pull request for me ? 
Thanks

File: soc/nxp/imxrt/Kconfig [hunk 1]
```diff
@@ -114,6 +114,24 @@ config DEVICE_CONFIGURATION_DATA
...
 
+config EXTERNAL_MEM_CONFIG_DATA
+	bool "External Memory Configuration Data"
+	help
+	  External memory configuration data (XMDC) provides an alternative
+	  configuration sequences which allows to intilialize the external memory
+	  at the boot time. This sequence allows to configure
+	  external memories (such as SDRAM) with more advanced option.
+	  This is a new alternative boot header compared to DCD, and DCD must be disabled
+	  in order to select this option.
```
Comment (dleach02) [inline soc/nxp/imxrt/Kconfig:125]: If DCD must be disabled when this option is enabled then shouldn't the DEVICE_CONFIG_DATA have a "depends on !EXTERNAL_MEM_CONFIG_DATA"?

**Review by danieldegrasse — COMMENTED** [review]
Why not switch the RT1170 EVK to use XCMD unconditionally? Is there a case where using the DCD would be preferred?

File: boards/nxp/mimxrt1170_evk/CMakeLists.txt [hunk 2]
```diff
@@ -28,17 +28,28 @@ if(CONFIG_NXP_IMXRT_BOOT_HEADER)
...
-    if(CONFIG_SRAM_BASE_ADDRESS EQUAL 0x80000000)
-      message(WARNING "You are using SDRAM as RAM but no device "
-        "configuration data (DCD) is included. This configuration may not boot")
-    endif()
+  elseif(CONFIG_EXTERNAL_MEM_CONFIG_DATA)
+    # Include external memory configuration data block for RT1170 EVK from NXP's HAL.
+    # This config is mutual exclusive with CONFIG_DEVICE_CONFIGURATION_DATA.
+    # because DCD and XMCD are overlapping.
+    zephyr_compile_definitions(XIP_BOOT_HEADER_XMCD_ENABLE=1)
+    zephyr_library_sources(${RT1170_BOARD_DIR}/xmcd/xmcd.c)
```
Comment (ngphibang) [inline boards/nxp/mimxrt1170_evk/CMakeLists.txt:52]: just to test zephyr reviewer mechanism

**Comment by trunghieulenxp** [comment]
Hi @danieldegrasse, 

The reason of not using the XMCD unconditionally is because the board RT1170 shared the same boot_header.ld and soc/nxp/imxrt/KConfig with other board such as RT10xx. The RT10xx variant does not support XMCD, so we cannot activate unconditionally XMCD. Moreover, the selection between DCD and XMCD is defined at this common KConfig, so we must add the selection between them.

File: soc/nxp/imxrt/Kconfig (hunk 1 above)
Comment (trunghieulenxp) [reply soc/nxp/imxrt/Kconfig:125]: Hi David, 
We cannot set DEVICE_CONFIG_DATA, "depends on !EXTERNAL_MEM_CONFIG_DATA" as this KConfig is shared between different platform like RT10xx.

**Comment by trunghieulenxp** [comment]
Hi @danieldegrasse,

Yes, we have changed the setting in the KConfig of RT1170 to only activate XMCD. As far as i know, there is no case when DCD would be preferred over XMCD for now, so i think it's ok to replace DCD with XMCD.

Thanks.

**Comment by danieldegrasse** [comment]
> Hi @danieldegrasse,

Yes, we have changed the setting in the KConfig of RT1170 to only activate XMCD. As far as i know, there is no case when DCD would be preferred over XMCD for now, so i think it's ok to replace DCD with XMCD.

Thanks.

sure- I should be more clear. I'm suggesting we just remove the DCD configuration code in the `CMakeLists.txt` file outright, since it is no longer needed

**Comment by trunghieulenxp** [comment]
Hi, I think that we should not completely remove DCD in the CMakefiles, because the selection between XMCD and DCD bootheader options still exists in the KConfig of imxrt folder.

**Comment by trunghieulenxp** [comment]
> > Hi @danieldegrasse,
> 
> Yes, we have changed the setting in the KConfig of RT1170 to only activate XMCD. As far as i know, there is no case when DCD would be preferred over XMCD for now, so i think it's ok to replace DCD with XMCD.
> 
> Thanks.
> 
> sure- I should be more clear. I'm suggesting we just remove the DCD configuration code in the `CMakeLists.txt` file outright, since it is no longer needed

Hi @danieldegrasse, sure, i've made the changes as you requested. Is it ok for you now? Thanks.

File: soc/nxp/imxrt/Kconfig (hunk 1 above)
Comment (danieldegrasse) [reply soc/nxp/imxrt/Kconfig:125]: You could set `EXTERNAL_MEM_CONFIG_DATA` to depend on `!DEVICE_CONFIGURATION_DATA` though, right?

File: soc/nxp/imxrt/Kconfig [hunk 3]
```diff
@@ -114,6 +114,24 @@ config DEVICE_CONFIGURATION_DATA
...
+
+config EXTERNAL_MEM_CONFIG_OFFSET
+	hex "External memory configuration offset"
+	depends on EXTERNAL_MEM_CONFIG_DATA
+	default 0x1040 if BOOT_FLEXSPI_NOR || BOOT_SEMC_NOR
+	help
+	  As specify by the boot ROM, the External Memory configuration data must be
```
Comment (danieldegrasse) [inline soc/nxp/imxrt/Kconfig:132]:
```suggestion
	  As specified by the boot ROM, the External Memory configuration data must be
```

File: soc/nxp/imxrt/Kconfig (hunk 1 above)
Comment (trunghieulenxp) [reply soc/nxp/imxrt/Kconfig:126]: Yes sure, i've modified as your suggestion. Thanks

File: soc/nxp/imxrt/Kconfig (hunk 3 above)
Comment (trunghieulenxp) [reply soc/nxp/imxrt/Kconfig]: I've changed with your suggestion. Thanks.

**Review by danieldegrasse — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]

**Pull request merged**